glyphs-reader/src/font.rs

```diff
@@ -347,359 +352,275 @@ struct NumberName {
+    fn as_fs_type(&self) -> Option<u16> {
+        Some(self.as_vec_of_ints()?.iter().map(|bit| 1 << bit).sum())
     }
 
-    fn unicode_range(&self) -> Option<&Vec<i64>> {
-        let Some(CustomParameterValue::UnicodeRange(bits)) = self.get("unicodeRanges") else {
-            return None;
-        };
-        Some(bits)
+    fn as_unicode_code_ranges(&self) -> Option<BTreeSet<u32>> {
+        let bits = self.as_vec_of_ints()?;
+        Some(bits.iter().map(|bit| *bit as u32).collect())
     }
 
-    fn codepage_range(&self) -> Option<&Vec<i64>> {
-        let Some(CustomParameterValue::CodepageRange(bits)) = self.get("codePageRanges") else {
-            return None;
-        };
-        Some(bits)
+    fn as_codepage_bits(&self) -> Option<BTreeSet<u32>> {
+        let bits = self.as_vec_of_ints()?;
+        bits.iter()
+            .map(|b| codepage_range_bit(*b as _))
+            .collect::<Result<_, _>>()
+            .ok()
     }
+}
 
-    fn panose(&self) -> Option<&Vec<i64>> {
+impl RawCustomParameters {
+    ////convert into the parsed params for a top-level font
+    fn cook(&self) -> Result<CustomParameters, Error> {
+        let mut params = CustomParameters::default();
+        let mut virtual_masters = Vec::<BTreeMap<String, OrderedFloat<f64>>>::new();
         // PANOSE custom parameter is accessible under a short name and a long name:
         //     https://github.com/googlefonts/glyphsLib/blob/050ef62c/Lib/glyphsLib/builder/custom_params.py#L322-L323
         // ...with the value under the short name taking precendence:
         //     https://github.com/googlefonts/glyphsLib/blob/050ef62c/Lib/glyphsLib/builder/custom_params.py#L258-L269
-        match self.get("panose").or_else(|| self.get("openTypeOS2Panose")) {
-            Some(CustomParameterValue::Panose(values)) => Some(values),
-            _ => None,
-        }
-    }
-}
-
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-enum CustomParameterValue {
-    Int(i64),
-    Float(OrderedFloat<f64>),
-    String(String),
-    Axes(Vec<Axis>),
-    AxesMappings(Vec<AxisMapping>),
-    AxisLocations(Vec<AxisLocation>),
-    GlyphOrder(Vec<SmolStr>),
-    VirtualMaster(Vec<AxisLocation>),
-    FsType(Vec<i64>),
-    UnicodeRange(Vec<i64>),
-    CodepageRange(Vec<i64>),
-    Panose(Vec<i64>),
-}
-
-/// Hand-parse these because they take multiple shapes
-impl FromPlist for RawCustomParameters {
-    fn parse(tokenizer: &mut Tokenizer<'_>) -> Result<Self, crate::plist::Error> {
-        use crate::plist::Error;
-        let mut params = Vec::new();
-
-        tokenizer.eat(b'(')?;
-
-        loop {
-            if tokenizer.eat(b')').is_ok() {
-                break;
-            }
-
-            tokenizer.eat(b'{')?;
+        let mut panose = None;
+        let mut panose_old = None;
+
+        for RawCustomParameterValue {
+            name,
+            value,
+            disabled,
+        } in &self.0
+        {
+            macro_rules! add_and_report_issues {
```

Comment:
Can we do this without a macro? If not, can we add a comment explaining why not?

Reply:
We need to use a macro because one of the inputs is the name of the field we're setting; can't pass that to a function. (Added a comment to this effect)